service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt

```diff
@@ -0,0 +1,163 @@
+                                                    'sentAt', msg.created,
+                                                    'title', msg.title,
+                                                    'content', msg.content,
+                                                    'senderName', msg.sender_name,
+                                                    'senderId', msg.sender_id))) AS messages
+         FROM threads
+                  JOIN messages msg ON msg.thread_id = threads.id
+         GROUP BY threads.count, threads.id, threads.type
+     )
+
+SELECT *
+FROM message_threads
+    """.trimIndent()
+
+    return this.createQuery(sql)
+        .bindMap(params)
+        .map(withCountMapper<MessageThread>())
+        .let(mapToPaged(pageSize))
+}
+
+fun Database.Read.getUnreadMessages(
+    accountIds: List<UUID>
+): Int {
+    // language=SQL
+    val sql = """
+        SELECT COUNT(DISTINCT m.id) AS count
+        FROM message_participants m
+        WHERE m.account_id = ANY(:accountIds) AND m.read_at IS NULL
+    """.trimIndent()
+
+    return this.createQuery(sql)
+        .bind("accountIds", accountIds)
+        .mapTo<Int>()
+        .first()
+}
+
+fun Database.Transaction.createMessageThread(
+    title: String,
+    content: String,
+    type: MessageType,
+    sender: MessageAccount,
+    recipientAccountIds: List<UUID>
+): UUID {
+    // language=SQL
+    val insertMessageSql = """
+        INSERT INTO message (title, content, sender_name)
+        VALUES (:title, :content, :senderName)        
+        RETURNING id
+    """.trimIndent()
+    val messageId = this.createQuery(insertMessageSql)
+        .bind("title", title)
+        .bind("content", content)
+        .bind("senderName", sender.name)
+        .mapTo<UUID>()
+        .first()
+
+    // language=SQL
+    val insertThreadSql = "INSERT INTO message_thread (message_type) VALUES (:messageType) RETURNING id"
+    val threadId = this.createQuery(insertThreadSql)
+        .bind("messageType", type)
+        .mapTo<UUID>()
+        .first()
+
+    // language=SQL
+    val insertMessageThreadMessagesSql = """
+        INSERT INTO message_thread_messages (message_id, thread_id) VALUES (:messageId, :threadId)
+    """.trimIndent()
+    this.createUpdate(insertMessageThreadMessagesSql)
+        .bind("messageId", messageId)
+        .bind("threadId", threadId)
+        .execute()
+
+    // language=SQL
+    val insertParticipantsSql = """
+         INSERT INTO message_participants (message_id, sender, account_id)
+         VALUES (:messageId, :sender, :accountId)
+    """.trimIndent()
+
+    val batch = this.prepareBatch(insertParticipantsSql)
+
+    val participants = listOf(Pair(sender.id, true)) + recipientAccountIds.map { Pair(it, false) }
+    for (p in participants) {
+        batch
+            .bind("messageId", messageId)
+            .bind("accountId", p.first)
+            .bind("sender", p.second)
+            .add()
+    }
```

Comment:
Nit: I would probably not define `participants` at all and just use `forEach` on the anonymous list on the right hand side of the definition on line 151.